SRS (#2)
NikitaKit1998/datakyt: docs/srs.md

# SOFTWARE REQUIREMENTS SPECIFICATION

## Prepared by

### Mykyta Kyt | Alex Zhytnyk

> 1. Introduction

Reviewer: Use links in the content or just remove it

Author: Okay

### 1.2 References

1. <https://www.wikiwand.com/en/Flask_(web_framework)> : Flask is a python based lightweight web application framework.
2. <https://www.postgresql.org/> : PostgreSQL is a powerful, open source object-relational database system.
3. <https://pypi.org/project/qrcode/>: Python's library for generating unique qr code for each device.
4. <https://cutt.ly/zf7MkUX> : This article is an explanation of how to connect PostgreSQL with Flask.
5. <https://hydra.cc/docs/intro/> : Hydra is an open-source Python framework that simplifies the development of research and other complex applications.

## 2\. Overall Description

### 2.1 Product perspective

An accounting database system stores the following information.

- **Equipment details:** It includes information about the equipment, such as laptops, monitors, mice, keyboards, headphones, which are issued to employees, as well as information about the equipment that is needed in other cities.

> - **Employee description:** It includes the employee's id, first name, last name, position, mail, phone number, birthday, city, project.

Reviewer: Why do we need a birthday?

Author: Because, in the employee table that I have, there is such a field.

Reviewer: This is not a valid reason to store unneeded info. Also let's keep personal info as less as possible

> - **Description of interaction:** It includes information which technique is given to each employee or information that the employee works on his technique. It also includes information about the workplace.

Reviewer: what about date when employee receives equipment? think about other missed fields.

Author: Okay

### 2.2 Product features

> This item presents the DFD-diagram, which shows the general cooperation in the system. They are presented in pictures 1.1-1.2 [![Picture 1.1 - DFD diagram A-0 level.](https://i.ibb.co/Jjn6rqM/Untitled-Document.png)](https://ibb.co/r2k8531)

Reviewer: Store images in the repo

Author: How to add it there?

Reviewer: Create a folder `images`, put all needed images there and commit.
In MD use relative path
```
![Image Name](images/image.png)
```

Picture 1.1 - DFD diagram A-0 level.

> [![Picture 1.2 - DFD diagram A0 level.](https://i.ibb.co/KDq4P58/DFD-A-0-1.png)](https://ibb.co/nBcdGwK)

Reviewer: Store images in the repo

Picture 1.2 - DFD diagram A0 level.

### 2.3 Operating environment

This product will work in any browser, as it will represent a web page. On PostgreSQL 9.5, Python 3.8\. This database will be located on Amazon AWS servers.

### 2.4 Design and implementation constraints

Our constraints are:

- Possibilities of Python programming language and PostgreSQL DBMS;

> - Writing Python code that conforms to PEP 8;

Reviewer: ```suggestion
- Writing Python code that satisfies PEP 8;
```

Author: Corrected

### 2.4 Design and implementation constraints

Our constraints are:

- Possibilities of Python programming language and PostgreSQL DBMS;
- Writing Python code that satisfies PEP 8;
- Constraints that may be imposed by the project business logic such as:

  - The ability to have multiple owners for one piece of equipment/furniture
  - The database is operated in two modes: full access and partial (read-only) access, etc.

  ## 3\. System features

  ### 3.1 Description and priority

- **Import/Export** to .csv file/ .xls file: saving or uploading tables.

- **Generating QR codes** with unique inventory number: unique number on the inventory, which will make it easier to count and keep records

- **Back up**: Saving the database before entering and changing information.

- **Data Structuring**: all information in a digital repository is organized into a clear hierarchical structure with records, tables or objects.
- **Multi-user Access**: provides multiple users access to all types of information stored in one and the same data store.

> - **Add/Remove** information about employee's or equipment in the database.

Reviewer 2: ```suggestion
- **Add/Remove** information about employees or equipment in the database.
```

Reviewer 3: Ok

> - **User-Friendly**: read-only access to the database.

Reviewer 2: Do you consider 'user-friendly' as a feature or as a quality?  Is it actually should appear in the Feature description section?

Reviewer 3: Ok, we will move it to other section

- **Reporting tools**: necessary data query
- **Inventory Monitoring**: the activity of checking stock.

## 4\. External interface requirements

> In my opinion, the application will intercommunicate directly through the web page. That is, to enter the site, login depending on the level of access (ordinary user, administrator), and continue to perform various operations, such as adding/removing employees or technicians, creating reports on the number of technicians or employees.

Reviewer 2: > the application will intercommunicate directly through the web page

With whom the application will intercommunicate directly? Did you mean a user or the database server?

> That is, to enter the site, login depending on the level of access (ordinary user, administrator)

Does it mean that every user will have to log in to be able to use the application or authorization will only be required for the admin role?

Reviewer 3: 1) Users will communicate with the database through website.
2)Two types of users: read-only or full access.

Author: Fixed

### 4.1 User interfaces

- Front - end software: Python 3.8 version
- Back - end software: PostgreSQL 9.5 version

### 4.2 Hardware interfaces

- a browser

> 

Reviewer 2: Could you please clarify why there are Python & Postgres in the 'user interfaces section' and browser in the hardware section as soon as it's python & Postgres which are not visible for user and browser which is visible? Shall it be vice versa? I'm not familiar with SRS terminology, but a browser in the hardware section looks weird to me, so I'd rather named it the 'software interfaces' or something (if it remains to be there)

Reviewer 3: Our Head Master said that it is correct.

Author: ![image](https://user-images.githubusercontent.com/59031709/95909975-0a18db00-0da8-11eb-8858-c8eb0c892e47.png)

## 5\. Non functional requirements

### 5.1 Software quality attributes

> For any system that is implemented in the software, several checks must be performed to minimize subsequent errors in its use. In our case we have used the following checks to detect errors:

Reviewer 2: ```suggestion
For any system that is implemented in the software, several checks must be performed to minimize subsequent errors in its usage. In our case we have used the following checks to detect errors:
```
Am I right, our guiding b2 expert? @aartiukh you taught me that word

- Writing Python code that satisfies PEP 8;
- Constraints that may be imposed by the project business logic such as:

  - The ability to have multiple owners for one piece of equipment/furniture
  - The database is operated in two modes: full access and partial (read-only) access, etc.

  ## 3\. System features

  ### 3.1 Description and priority

- **Import/Export** to .csv file/ .xls file: saving or uploading tables.

- **Generating QR codes** with unique inventory number: unique number on the inventory, which will make it easier to count and keep records

- **Back up**: Saving the database before entering and changing information.

- **Data Structuring**: all information in a digital repository is organized into a clear hierarchical structure with records, tables or objects.

- **Multi-user Access**: provides multiple users access to all types of information stored in one and the same data store.

- **Add/Remove** information about employees or equipment in the database.

- **Reporting tools**: necessary data query
- **Inventory Monitoring**: the activity of checking stock.

## 4\. External interface requirements

> In my opinion, the application will intercommunicate directly through the web page with users. That is, to enter the site, login depending on the level of access (ordinary user, administrator this means that there will be an account for the admin and the ordinary user), and continue to perform various operations, such as adding/removing employees or technicians, creating reports on the number of technicians or employees.

Reviewer 2: ```suggestion
In my opinion, the application will intercommunicate directly through the web page with users. That is, to enter the site, login is required. Depending on the level of access (ordinary user or administrator -- this means that there will be an account for the admin and the ordinary user), the user continues to perform various operations, such as adding/removing employees or technicians, creating reports on the number of technicians or employees.
```
